File: Java_ChildrenCompetition/src/main/java/service/ParticipantService.java
package service;

import exception.InvalidUsernameException;
import exception.TestJoinedException;
import exception.TestLimitException;
import model.Participant;
import model.Test;
import repository.IParticipantRepository;
import repository.ITestRepository;

import java.util.List;

public class ParticipantService implements IParticipantService {
    public final IParticipantRepository participantRepository;
    public final ITestRepository testRepository;

    public ParticipantService(IParticipantRepository participantRepository, ITestRepository testRepository) {
        this.participantRepository = participantRepository;
        this.testRepository = testRepository;
    }

    @Override
    public void save(String username, String name, int age, int testId) {
        Participant participant = participantRepository.findByUsername(username);
        if(participant == null){
            Participant newParticipant = new Participant(username, name, age);
            participantRepository.save(newParticipant);
            return;
        }
        if(participant.getAge() != age || !participant.getName().equals(name)){
            throw new InvalidUsernameException();
        }

        List<Test> testList = (List<Test>) testRepository.findAllTestsForParticipant(participant.getId());
        if(testList.size() >= 2){
            throw new TestLimitException();
        }

        if(testList.get(0).getId().equals(testId)){
            throw new TestJoinedException();
        }
    }

    @Override
    public Participant findByUsername(String username) {
        return participantRepository.findByUsername(username);
    }


    @Override
    public List<Participant> findAllParticipantsForTest(Integer id) {
        return (List<Participant>) participantRepository.findAllParticipantsForTest(id);
    }
}

File: CSharp_ChildrenCompetitionGUI/CSharp_ChildrenCompetitionGUI/CSharp_ChildrenCompetitionGUI/service/ParticipantService.cs
using System.Collections.Generic;
using CSharp_ChildrenCompetition.repository;
using CSharp_ChildrenCompetitionGUI.model;
using CSharp_ChildrenCompetitionGUI.repository;

namespace CSharp_ChildrenCompetitionGUI.service
{
    public class ParticipantService : IParticipantService<int, Participant>
    {
        public IParticipantRepository<int, Participant> participantRepository;
        public ITestRepository<int, Test> testRepository;

        public ParticipantService(IParticipantRepository<int, Participant> participantRepository, ITestRepository<int, Test> testRepository)
        {
            this.participantRepository = participantRepository;
            this.testRepository = testRepository;
        }

        public void save(string username, string name, int age, int testId)
        {
            // throw new System.NotImplementedException();
            Participant participant = participantRepository.findByUsername(username);
            if (participant == null)
            {
                Participant newParticipant = new Participant(username, name, age);
                participantRepository.save(newParticipant);
                return;
            }

            if (participant.age != age || !participant.name.Equals(name))
            {
                throw new InvalidUsernameException();
            }

            List<Test> testList = (List<Test>)testRepository.findAllTestsForParticipant(participant.id);
            if (testList.Count >= 2)
            {
                throw new TestLimitException();
            }

            if (testList[0].id.Equals(testId))
            {
                throw new TestJoinedException();
            }
        }

        public Participant findByUsername(string username)
        {
            // throw new System.NotImplementedException();
            return participantRepository.findByUsername(username);
        }

        public List<Participant> findAllParticipantsForTest(int id)
        {
            // throw new System.NotImplementedException();
            return (List<Participant>) participantRepository.findAllParticipantsForTest(id);
        }
    }
}